Initial setup of the TailwindCSS hugo module

ringods opened the pull request:
Hello @davidsneighbour,

I started this draft pull request for 2 purposes:

1. are you interested in accepting TailwindCSS as an additional styling library next to Bootstrap.
2. as a base to discuss the questions I have

Hoping that you are positive on 1, I need to know how everything needs to fit in your release procedure driven by your configuration libraries.

Contrary to the setup of your `bootstrap5` module, I can't use the upstream TailwindCSS git repo as a dependency in this module. I need the built version of TailwindCSS coming from the npm package. I added `tailwindcss` as a dependency in my `package.json` and added a `build` script which shows the intent: it copies the content from `./node_modules/tailwindcss` to `assets/tailwindcss`. After that, I still need to commit the content in `assets/tailwindcss` before releasing the module. The content of the assets folder needs to be updated with the contents of the corresponding npm package.

How should I integrate the following steps into the release proces:

```
$ rm -rf ./assets/tailwindcss
$ cp `./node_modules/tailwindcss` to `./assets/tailwindcss`
$ git add .
$ git commit -m "<appropriate commit message here>"
```

Hoping for a positive message. 😉 

Signed-off-by: Ringo De Smet <ringo@de-smet.name>

davidsneighbour commented:
Hi @ringods! Thank you for this pull request. Of course, I am interested in adding Tailwind to the libraries setup. Due to its complexity, however, I think it might be better located in the @dnb-org/configurations "vicinity". 

The following is a quick walkthrough for my own tailwind based setup. If you want I can give you access to a site in progress where I am using tailwind. 

First of all, we don't need to copy tailwind around. With some work (webpack to the rescue) it will work just with an npm install. What I did not get to work though is that new JIT feature. I also think it's not required for a Hugo dev setup, so I did not investigate that further.

It's a bit complicated, so I think it's best you have a look at the website repo, let me know if you want to access it, it's private currently. Long story short: You set up tailwind as you would normally and add webpack. Then you run webpack once (to create the first compiled style.css), then you run webpack and Hugo in watch mode in parallel. When tailwind changes a file, Hugo processes it. 

Most of the samples below are probably overkill, but better to add too much than non-working stuff.

1) install dependencies in package.json (dnb-org/all is a lazy solution to include all my configs, you probably need only dnb-org/tools to have tools to run npm parallel and rimraf to delete logs and caches)

```
"dependencies": {
    "@dnb-org/all": "3.3.5",
    "@tailwindcss/aspect-ratio": "0.3.0",
    "@tailwindcss/forms": "0.3.4",
    "@tailwindcss/typography": "0.4.1",
    "alpinejs": "3.4.2",
    "lazysizes": "5.3.2",
    "lite-youtube-embed": "^0.2.0",
    "pixrem": "5.0.0",
    "precss": "4.0.0",
    "tailwindcss": "2.2.17"
  },
```

2) server scripts in package.json

```
"scripts": {
    "build": "run-s build:critical && build:hugo",
    "build:critical": "./bin/build-critical-css.sh",
    "build:highlights": "hugo gen chromastyles --style=native > assets/css/utilities/syntax.css",
    "build:hugo": "hugo --gc --minify",
    "build:new": "npm run build:webpack && npm run build:hugo",
    "build:webpack": "cross-env NODE_ENV=production webpack --config webpack.production.js",

    "serve": "npm run build:webpack && npm-run-all -p serve:**",
    "serve:hugo": "rimraf hugo.log && hugo server --bind 192.168.1.201 --baseURL http://192.168.1.201 --disableFastRender --i18n-warnings --navigateToChanged  --templateMetrics --templateMetricsHints --verbose --verboseLog --path-warnings --buildDrafts --buildExpired --buildFuture --watch --enableGitInfo --forceSyncStatic --log true --logFile hugo.log -v",
    "serve:webpack": "webpack --watch --config webpack.development.js",
  },
```

3) webpack config in webpack.development.js

```
// eslint-disable-next-line import/no-extraneous-dependencies
const { merge } = require("webpack-merge");
// eslint-disable-next-line import/extensions
const common = require("./webpack.common.js");

module.exports = merge(common, {
  mode: "production",
});
```

4) common webpack config in webpack.common.js

```
module.exports = {
  target: ["browserslist"],

  entry: {
    main: path.join(__dirname, "assets/js", "theme.js"),
  },

  output: {
    path: path.join(__dirname, "static/assets/"),
    filename: "[name].js",
    chunkFilename: "[id].css",
    clean: true,
  },

  performance: {
    maxEntrypointSize: 400000,
    maxAssetSize: 100000,
    hints: "warning",
  },

  module: {
    rules: [
      {
        loader: "babel-loader",
        test: /\.js?$/,
        exclude: /node_modules/,
      },
    ],
  },
};
```

5) styles setup in assets/css/styles.css etc

```
@tailwind base;
@layer base {
  /*noinspection CssInvalidImport*/
  @import "base/_index.css";
}

@tailwind components;
@layer components {
  /*noinspection CssInvalidImport*/
  @import "components/_index.css";
}

@tailwind utilities;
@layer utilities {
  /*noinspection CssInvalidImport*/
  @import "utilities/_index.css";
}
```

6) style tag in the layout

```
{{- $options := dict "inlineImports" true -}}
{{- $styles := resources.Get "css/styles.css" -}}
{{- $styles = $styles | resources.PostCSS $options -}}
{{- if hugo.IsProduction -}}
  {{- $styles = $styles | minify | fingerprint | resources.PostProcess -}}
{{- end -}}
<link href="{{- $styles.Permalink -}}" rel="stylesheet" />
```

Looking at this, it might be double work. Not sure about the PostCss part here. I did not check the process for a while, but running a quick update just now and then the server the site came up ok, so it's not a destructive thing, just wrong maybe. The main idea is that hugo takes the style.css from the static directory on updates, but the layout file processes the tailwind styles with postcss. webpack might not be needed. hmm.

7) tailwind config in tailwind.config.js

```
const typography = require("@tailwindcss/typography");
const forms = require("@tailwindcss/forms");
const colors = require("tailwindcss/colors");
const aspectRatio = require("@tailwindcss/aspect-ratio");

module.exports = {
  // See https://tailwindcss.com/docs/configuration#important
  important: false,
  // prefix: "dnb-",
  purge: {
    enabled: process.env.HUGO_ENVIRONMENT === "production",
    content: [
      "./hugo_stats.json",
      "./layouts/**/*.{vue,js,ts,jsx,tsx,html}",
      "./assets/**/*.{vue,js,ts,jsx,tsx,html}",
    ],
    extractors: [
      {
        extractor: (content) => {
          // noinspection JSUnresolvedVariable
          const els = JSON.parse(content).htmlElements;
          return els.tags.concat(els.classes, els.ids);
        },
        extensions: ["json"],
      },
    ],
    // in case of missing styles read this part of the documentation
    // https://tailwindcss.com/docs/optimizing-for-production#removing-all-unused-styles
    mode: "all",
    preserveHtmlElements: false,
    safelist: ["text-center"],
    // https://purgecss.com/configuration#options
    options: {
      // https://tailwindcss.com/docs/optimizing-for-production#removing-unused-keyframes
      keyframes: true,
      fontFace: true,
      variables: true,
      rejected: false,
    },
  },
  // https://tailwindcss.com/docs/dark-mode
  darkMode: false, // or 'media' or 'class'
  plugins: [forms, aspectRatio, typography],
  
};
```

davidsneighbour commented:
Hmm, I guess that JIT feature needs to be added, if its going to be the default in v3.

https://www.themes.dev/blog/tailwind-css-version-3/

ringods commented:
@davidsneighbour contrary to your npm+webpack setup, I wanted to go a completely other route: build a Hugo module and only rely on [Hugo Pipes](https://gohugo.io/hugo-pipes/). 

Am I right, that if this library would rely on npm+webpack, that this requirement is eventually pushed onto every user as well? That is not what I had in mind or wanted myself.

In the Hugo module setup I had in mind, the only requirement a user needs to have is a [globally installed PostCSS](https://gohugo.io/hugo-pipes/postcss/), but that is a requirement of Hugo Pipes for CSS processing in general.

The closest I had in mind was this as a counterpart: https://github.com/dirkolbrich/hugo-theme-tailwindcss-starter

I only don't want this module to be a theme. I had in mind that this library, only offers the raw TailwindCSS as a Hugo module, added by another Hugo module offering the integration of TailwindCSS with Hugo Pipes. Both together could be integrated as Hugo modules in a theme by anyone.

Your clean separation into separate modules within dnb-org was my trigger here.

What do you think about the opposite route I want to take?

davidsneighbour commented:
I like that route, but please look at my style.html sample. it looks like I am doing that already. I did not check the website for some months, so I am not 100% sure. The proper way is of course to put it into the pipes and let postcss do it's thing. I will test my implementation later today to see. It looks to me as if I was working on moving it in without removing the external implementation yet. It's for my own website, so it has less importance than other projects ;)

I'll add you to the repo so you can check it out. 

And I am pretty sure you are wrong about the global PostCSS installation. I do all my projects with local install. Two points here: don't use Hugo snap install and always install all plugins locally too.

davidsneighbour commented:
The main issue I am seeing for both ways is, that the new JIT feature is currently not working on both approaches.

ringods commented:
@davidsneighbour I think we are getting somewhere. Your site setup is indeed leveraging Hugo Pipes with PostCSS processing. As you mentioned, I think `webpack` is indeed doing the same and can be removed.

This is what I did in new commits:

* I no longer copy `tailwindcss` into this module, but I added a `package.hugo.json` file. This requires Hugo 0.75.0 as a minimum version. See the article referenced below.
* I dropped `rimraf` and `copyfiles` dev dependencies

When this module is included in a higher level project, the user there must run `hugo mod npm pack` to generate a global `package.json` file including the npm requirements from all used modules, including this one.

References:
* https://gohugo.io/news/0.75.0-relnotes/#npm-pack

davidsneighbour commented on tailwindcss/config/_default/module.toml:
```diff
@@ -0,0 +1,7 @@
+[hugoVersion]
+extended = true
+min = "0.75.0"
```
can we go up to 0.84.0? that one introduced the config revamp that IMHO has the `_merge` in params.toml later on.

davidsneighbour replied:
Can you please make the bootstrap change a separate pull request? I am not 100% sure, but having it in this PR will lead to the Bootstrap changelog having tailwind commits in it (I will be doing squash merge).

davidsneighbour replied:
After that I'll merge tonight.

ringods replied:
@davidsneighbour reverted the bootstrap5 min version.

davidsneighbour commented:
Hi @ringods, would you be ok with me merging this into a branch `feat/tailwindcss` instead of main so that we can add more commits before adding a 1.0 release? This way we both can add commits and when it's ready for "out in the wild" I am merging it with all commits in that branch. If not, that's ok. we leave it on a 0.0-version number until it's finished.

davidsneighbour commented:
I just merged it ;) Let's continue versus the feat/tailwindcss branch.

davidsneighbour commented:
see https://github.com/dnb-org/libraries/tree/feat/tailwindcss

ringods commented:
@davidsneighbour did you mixup the branches? I see `tailwindcss` folder already on `main`, but your branch `feat/tailwindcss` **doesn't** contain tailwind...

ringods commented:
@davidsneighbour, you wrote in a previous comment:

> so that we can add more commits before adding a 1.0 release?

What other commits did you have in mind? Do you want the rest of the setup required to bridge the gap between the core TailwindCSS package and the Hugo Pipes setup also in this library module? Or should I add the "integration" part as another module in e.g. your components repo?
